### PR TITLE
UCT: API change for non-contiguous datatype support

### PR DESCRIPTION
**Pull request by alex--m**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4054/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4057/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4069/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4079/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
Can I return more than a single key pre region ? It is not very clear from API proposal

File: src/uct/api/uct.h [hunk 1]
```diff
@@ -322,8 +325,9 @@ enum {
                                               remote memory key for remote memory
                                               operations */
     UCT_MD_FLAG_ADVISE    = UCS_BIT(4),  /**< MD support memory advice */
-    UCT_MD_FLAG_FIXED     = UCS_BIT(5)   /**< MD support memory allocation with
+    UCT_MD_FLAG_FIXED     = UCS_BIT(5),  /**< MD support memory allocation with
                                               fixed address */
+    UCT_MD_FLAG_REG_NC    = UCS_BIT(6)   /**< MD support non-contig registration */
```
Comment (yosefe) [inline src/uct/api/uct.h]: why need a flag in memory domain, since no operation/function is added to the memory domain?

File: src/uct/api/uct.h [hunk 2]
```diff
@@ -337,8 +341,9 @@ enum uct_md_mem_flags {
                                                 mapping may be deferred until
                                                 it is accessed by the CPU or a
                                                 transport. */
-    UCT_MD_MEM_FLAG_FIXED    = UCS_BIT(1)  /**< Place the mapping at exactly
+    UCT_MD_MEM_FLAG_FIXED    = UCS_BIT(1), /**< Place the mapping at exactly
                                                 defined address */
+    UCT_MD_MEM_FLAG_EMPTY    = UCS_BIT(2)  /**< Create empty handle (for UMR) */
```
Comment (yosefe) [inline src/uct/api/uct.h]: how should it be used?
also UMR is IB-specific terminology which can't be used it uct common code

File: src/uct/api/uct.h [hunk 3]
```diff
@@ -1603,6 +1608,35 @@ UCT_INLINE_API ucs_status_t uct_ep_am_zcopy(uct_ep_h ep, uint8_t id, void *heade
     return ep->iface->ops.ep_am_zcopy(ep, id, header, header_length, iov, iovcnt, comp);
 }
 
+
+/**
+ * @ingroup UCT_AM
+ * @brief Register non-contiguous memory.
```
Comment (yosefe) [inline src/uct/api/uct.h]: Please add more description about this function, including:
 - how different types of memory should be described
 - completion semantics

File: src/uct/api/uct.h [hunk 4]
```diff
@@ -1603,6 +1608,35 @@ UCT_INLINE_API ucs_status_t uct_ep_am_zcopy(uct_ep_h ep, uint8_t id, void *heade
...
+ *
+ * @param [in]  ep           Destination endpoint handle.
+ * @param [in]  iov          Points to an array of @ref ::uct_iov_t structures.
+ *                           The @a iov pointer must be valid address of an array
+ *                           of @ref ::uct_iov_t structures. A particular structure
+ *                           pointer must be valid address. NULL terminated pointer
+ *                           is not required.
+ * @param [in]  iovcnt       Size of the @a iov data @ref ::uct_iov_t structures
+ *                           array. If @a iovcnt is zero, the data is considered empty.
+ *                           @a iovcnt is limited by @ref uct_iface_attr_cap_am_max_iov
```
Comment (yosefe) [inline src/uct/api/uct.h]: uct_iface_attr_cap_am_max_iov seems inapplicable - need to introduce mem_reg.max_iov?

File: src/uct/api/uct_def.h [hunk 5]
```diff
@@ -137,11 +137,12 @@ typedef int                      uct_worker_cb_id_t;
  *
  */
 typedef struct uct_iov {
-    void     *buffer;   /**< Data buffer */
+    void     *buffer;   /**< Data buffer (if null - stride/count reference previous) */
```
Comment (yosefe) [inline src/uct/api/uct_def.h]: "stride/count reference previous" - need to elaborate

File: src/uct/api/uct_def.h [hunk 6]
```diff
@@ -137,11 +137,12 @@ typedef int                      uct_worker_cb_id_t;
...
     size_t    length;   /**< Length of the payload in bytes */
     uct_mem_h memh;     /**< Local memory key descriptor for the data */
     size_t    stride;   /**< Stride between beginnings of payload elements in
                              the buffer in bytes */
+    size_t    ilv_ratio;/**< Interleaving ratio - Only if enabled on first iov */
```
Comment (yosefe) [inline src/uct/api/uct_def.h]: need to elaborate how interleaving works

File: src/uct/api/tl.h [hunk 7]
```diff
@@ -163,6 +163,12 @@ typedef struct uct_iface_ops {
                                      const uct_device_addr_t *dev_addr,
                                      const uct_ep_addr_t *ep_addr);
 
+    /* endpoint - memory registration */
+
+    ucs_status_t (*ep_mem_reg_nc)(uct_ep_h ep, const uct_iov_t *iov,
```
Comment (yosefe) [inline src/uct/api/tl.h]: not sure i would use the terminology "noncontig" since the memory _can_ be contiguous. maybe just use "ep_mem_reg"? we can decide that memory registration on an endpoint is more "generic" than over an interface
@shamisp what do you think?

**Comment by alex--m** [comment]
@yosefe fixed according to feedback, please review

File: src/uct/api/tl.h [hunk 8]
```diff
@@ -165,6 +165,12 @@ typedef struct uct_iface_ops {
                                      const uct_device_addr_t *dev_addr,
                                      const uct_ep_addr_t *ep_addr);
 
+    /* endpoint - memory registration */
+
+    ucs_status_t (*ep_mem_reg)(uct_ep_h ep, const uct_iov_t *iov,
+                               size_t iovcnt, uct_md_h *md_p,
```
Comment (yosefe) [inline src/uct/api/tl.h]: should be `uct_md_h md`, md is input param

File: src/uct/api/uct.h [hunk 9]
```diff
@@ -459,6 +462,17 @@ struct uct_iface_attr {
             } rndv;                      /**< Attributes related to rendezvous protocol */
         } tag;                           /**< Attributes for TAG operations */
 
+        struct {
+            size_t           max_iov;    /**< Maximal @a iovcnt parameter in
+                                              @ref uct_ep_mem_reg */
+            size_t           max_dim;    /**< Maximal dimension for a strided memory
+                                              layout with @ref uct_iov_t */
+            size_t           max_short;  /**< Maximal @a iovcnt parameter in
```
Comment (yosefe) [inline src/uct/api/uct.h]: why is this called 'max_short'?

File: src/uct/api/uct.h [hunk 10]
```diff
@@ -459,6 +462,17 @@ struct uct_iface_attr {
...
+                                              @ref uct_ep_mem_reg to fit in a single
+                                              key (for latency optimization) */
+            size_t           max_nested; /**< Maximal nesting level of memory keys */
```
Comment (yosefe) [inline src/uct/api/uct.h]: s/max_nested/max_indirect/

File: src/uct/api/uct.h [hunk 11]
```diff
@@ -1655,6 +1669,54 @@ UCT_INLINE_API ucs_status_t uct_ep_am_zcopy(uct_ep_h ep, uint8_t id,
...
+ * The layout is composed of one or more @ref ::uct_iov_t objects. For example,
+ * a simple contiguous buffer would be a single object containing a pointer to
+ * the buffer and its length (count is 1, and the rest of the fields are zero).
+ * For a complex memory layout, such as two interleaved vectors, the function
+ * should accept two objects, one for each vector, each set with the stride between
+ * two items, the length of an item, their amount and the interleaving ratio:
+ * how many of items from a vector to put before moving to the next vector (for
+ * a pattern of AABAABAAB... the interleaving ratio is 2 for vector A and 1 for B).
+ *
+ * The user must destroy the handle by calling @ref uct_md_mem_free with @a md_p ,
```
Comment (yosefe) [inline src/uct/api/uct.h]: s/md_p/md

File: src/uct/api/uct.h (hunk 4 above)
Comment (yosefe) [inline src/uct/api/uct.h]: pls address this comment

File: src/uct/api/uct.h [hunk 12]
```diff
@@ -1655,6 +1669,54 @@ UCT_INLINE_API ucs_status_t uct_ep_am_zcopy(uct_ep_h ep, uint8_t id,
...
+ * @param [in]  iov          Points to an array of @ref ::uct_iov_t structures.
+ *                           The @a iov pointer must be valid address of an array
+ *                           of @ref ::uct_iov_t structures. A particular structure
+ *                           pointer must be valid address. NULL terminated pointer
+ *                           is not required.
+ * @param [in]  iovcnt       Size of the @a iov data @ref ::uct_iov_t structures
+ *                           array. If @a iovcnt is zero, the data is considered empty.
+ *                           @a iovcnt is limited by @ref uct_iface_attr_cap_am_max_iov
+ *                           "uct_iface_attr::cap::mem::max_iov"
+ * @param [out] md_p         Filled with the memory domain handle, for destruction.
```
Comment (yosefe) [inline src/uct/api/uct.h]: why is memory domain an output parameter?
why do we even need it, we already have endpoint parameter which is on a specific memory domain?

File: src/uct/api/uct_def.h [hunk 13]
```diff
@@ -99,7 +99,13 @@ typedef int                      uct_worker_cb_id_t;
  * @brief Structure for scatter-gather I/O.
  *
  * Specifies a list of buffers which can be used within a single data transfer
- * function call.
+ * function call, or to be registered as a memory layout with @ref uct_ep_mem_reg .
+ *
+ * A single entry of this data structure represents either a contiguous array of
+ * items or a simple strided vector of items. An X-dimensional strided vector would
+ * consist of X consecutive structures, where only the first has a non-zero buffer
+ * pointer, and each contains the stride and count of the i-th dimension.
```
Comment (yosefe) [inline src/uct/api/uct_def.h]: seems counter-intuitive to require the user to pass a "serialized tree". Why not use the memory key to describe a hierarchy?

File: src/uct/api/uct_def.h [hunk 14]
```diff
@@ -111,6 +117,13 @@ typedef int                      uct_worker_cb_id_t;
     |<---- stride ----->|<---- stride ----->|
    @endverbatim
  *
+ * In order to describe an interleaved memory layout, each structure contains
```
Comment (yosefe) [inline src/uct/api/uct_def.h]: Can't the same be achieved with "count" field?

File: src/uct/api/tl.h (hunk 8 above)
Comment (alex--m) [inline src/uct/api/tl.h]: it's an output param, so that the user knows how to call uct_md_mem_free() on the resulting memh. Otherwise i'm not sure the user has a way to know which md is used.

File: src/uct/api/uct.h (hunk 9 above)
Comment (alex--m) [inline src/uct/api/uct.h]: inspired by "short" messages - up to some X iov-s UMR can post inline, and for >X latency is worse. the function will work either way, but UCP wants to act accordingly.

File: src/uct/api/uct.h (hunk 10 above)
Comment (alex--m) [inline src/uct/api/uct.h]: ok

File: src/uct/api/uct.h (hunk 12 above)
Comment (alex--m) [inline src/uct/api/uct.h]: how the user can go from ep to md?

File: src/uct/api/uct_def.h (hunk 13 above)
Comment (alex--m) [inline src/uct/api/uct_def.h]: presumably overhead. I don't mind limiting one memh to one dimension, but i suspect 2/3D will be common enough (will work, but probably some penalty).

File: src/uct/api/uct_def.h (hunk 14 above)
Comment (alex--m) [inline src/uct/api/uct_def.h]: no, because count tells you how many items. For example: interleaving "A_A_A_A_" and "B_B_B_B_" (count is 4 and stride is 2) can be ABABABAB for 1:1 ratio, or AABBAABB for 2:2 ratio, or ABBABB for 1:2 ratio.

File: src/uct/api/uct.h (hunk 11 above)
Comment (alex--m) [inline src/uct/api/uct.h]: PSB

File: src/uct/api/tl.h (hunk 8 above)
Comment (yosefe) [inline src/uct/api/tl.h]: there is no need for this param because an ep is already created on a particular iface which was created on a particular md. so the md should be known to the user up-front.

File: src/uct/api/uct.h (hunk 9 above)
Comment (yosefe) [inline src/uct/api/uct.h]: seems like need to expose performance estimations for UMR overhead/latency from UCT, since max_short does not give an idea on what the latency actually is. Can't make the decision of pack/unpack vs. UMR, for example.

File: src/uct/api/uct.h (hunk 12 above)
Comment (yosefe) [inline src/uct/api/uct.h]: see above, the md should be known up-front

File: src/uct/api/uct_def.h (hunk 13 above)
Comment (yosefe) [inline src/uct/api/uct_def.h]: a single pointer dereference overhead should not be significant for zero-copy APIs, on the other hand the usage (and corresponding UCP code) should be simpler

File: src/uct/api/uct_def.h (hunk 14 above)
Comment (yosefe) [inline src/uct/api/uct_def.h]: can we encode the latter example as:
dtype C {
  A_ (count:1)
  B_ (count:2)
}
send(buffer, dtype=C, count=2)

File: src/uct/api/uct.h (hunk 4 above)
Comment (alex--m) [inline src/uct/api/uct.h]: I added uct_iface_attr.cap.mem.max_iov and changed the documentation for ep_mem_reg(), but not the other functions (ep_send_*) still using uct_iface_attr.cap.am.max_iov. Is that ok?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4256/ for details (Mellanox internal link).

File: src/uct/api/uct.h (hunk 4 above)
Comment (yosefe) [inline src/uct/api/uct.h]: but this is the documentation of `uct_ep_mem_reg`, it should use uct_iface_attr_cap_**mem**_max_iov

File: src/uct/api/uct_def.h (hunk 14 above)
Comment (alex--m) [inline src/uct/api/uct_def.h]: I don't see how, because it's ambiguous. For each letter you need to store both count (how many items), length (per-itme) and pattern (ilv_ratio). For example, how would you encode the difference between ABABABAB and AABBAABB ?

Comment (yosefe) [inline src/uct/api/uct_def.h]: ABABABAB:
dtype C {
A_ (count:1)
B_ (count:1)
}
send(buffer, dtype=C, count=4)

AABBAABB:
dtype C {
A_ (count:2)
B_ (count:2)
}
send(buffer, dtype=C, count=2)

Comment (alex--m) [inline src/uct/api/uct_def.h]: In that case - what's the difference between ABABABAB and ABAB ?

Comment (alex--m) [inline src/uct/api/uct_def.h]: Correction: where do you state you have just 2 types, A & B, and not A & B & C ? I see you're using count for the length, rather than the number of dtype descriptions.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4264/ for details (Mellanox internal link).

File: .gitignore [hunk 15]
```diff
@@ -63,3 +63,4 @@ tags
 .pydevproject
 /ucx-*.*.*
 /v*.*.*
+/install-release/
```
Comment (shamisp) [inline .gitignore]: should not be part of this PR

File: src/uct/api/uct_def.h [hunk 16]
```diff
@@ -123,11 +136,15 @@ typedef int                      uct_worker_cb_id_t;
...
-    void     *buffer;   /**< Data buffer */
+    void     *buffer;   /**< Data buffer (if null - stride/count reference previous) */
     size_t    length;   /**< Length of the payload in bytes */
     uct_mem_h memh;     /**< Local memory key descriptor for the data */
     size_t    stride;   /**< Stride between beginnings of payload elements in
                              the buffer in bytes */
+    size_t    ilv_ratio;/**< Interleaving ratio - item count for this vector.
+                             This applies only to strided types, where stride is
+                             nonzero, and only if all the ratios are nonzero
+                             on the same function invocation. */
```
Comment (shamisp) [inline src/uct/api/uct_def.h]: I want to see diagram added describing this concepts.

File: .gitignore (hunk 15 above)
Comment (alex--m) [inline .gitignore]: my bad, accidental :)

File: src/uct/api/uct_def.h (hunk 16 above)
Comment (alex--m) [inline src/uct/api/uct_def.h]: diagram added.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4269/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4270/ for details (Mellanox internal link).

File: src/uct/api/uct_def.h (hunk 14 above)
Comment (yosefe) [inline src/uct/api/uct_def.h]: i've used UCP notation before, i'll try to be more clear how to do this in UCT:
ep_mem_reg { count=1, &memh_A}
ep_mem_reg { count=1, &memh_B}
ep_mem_reg { {memh_A, count=1}, {memh_B, count=1}, &memh_AB}
ep_mem_reg { {memh_AB, count=4},  &memh_ABABABAB}
ep_mem_reg { {memh_AB, count=2},  &memh_ABAB}

Comment (alex--m) [inline src/uct/api/uct_def.h]: There are problems with this proposal:
1. memh_ABAB exceeds the bounds of its underlying memh_AB, which makes bounds check complicated & costly (recursive down) for what is supposed to be practically data-path. 
2. This hurts the performance of UMR since you need more calls to ep_mem_reg before you can start sending
3. It complicates the user that now needs more memh-s for the same pattern
4. The creation of the UMR also takes more time because you need to combine multiple datatype structs to post the WR (the WR for memh_ABAB will not point to memh_AB - it needs to point to memh_A and memh_B directly, otherwise we have overhead on the HCA filling each send request)

The only benefit I see to this is removing the "extra" interleaving ratio field from the API... Is it worth it?

**Comment by alex--m** [comment]
@yosefe is this still waiting on author response? if yes - which question?

File: src/uct/api/uct_def.h (hunk 14 above)
Comment (yosefe) [inline src/uct/api/uct_def.h]: I want to see how we can make the simplest possible API,
we can also add "repeat_count" to ep_mem_reg, and drop "count" and "ilv_ratio" from uct_iov_t (keep only buffer, key, len, stride)
Therefore, ABBBCCABBBCCABBBCC could be represented as follows:
```
iov[0].ptr = &A
iov[0].length=1
iov[0].memh= key_A
iov[1].ptr = &B
iov[1].length=3
iov[1].memh= key_B
iov[2].ptr = &C
iov[2].length=2
iov[2].memh= key_C
ep_mem_reg(iov, iovcnt=3, repeat_count=3)
```
